source/darwin/nw_socket.c

```diff
@@ -1039,82 +1035,121 @@ static void s_process_listener_success_task(struct aws_task *task, void *args, e
+        struct nw_socket *new_nw_socket = new_socket->impl;
+        new_nw_socket->os_handle.nw_connection = task_args->new_connection;
+        new_nw_socket->connection_setup = true;
 
-                nw_connection_set_state_changed_handler(
-                    new_socket->io_handle.data.handle, ^(nw_connection_state_t state, nw_error_t error) {
-                      s_schedule_connection_state_changed_fn(
-                          new_socket, new_nw_socket, new_nw_socket->os_handle.nw_connection, state, error);
-                    });
+        // Setup socket state to start read/write operations.
+        s_set_socket_state(new_nw_socket, new_socket, CONNECTED_READ | CONNECTED_WRITE);
 
-                // released when the connection state changed to nw_connection_state_cancelled
-                s_socket_acquire_internal_ref(new_nw_socket);
+        nw_connection_set_state_changed_handler(
+            new_socket->io_handle.data.handle, ^(nw_connection_state_t state, nw_error_t error) {
+              int nw_error_code = error ? nw_error_get_error_code(error) : 0;
+              int crt_error_code = nw_error_code ? s_determine_socket_error(nw_error_code) : AWS_OP_SUCCESS;
+              AWS_LOGF_TRACE(
+                  AWS_LS_IO_SOCKET,
+                  "id=%p handle=%p: nw_connection_set_state_changed_handler invoked with nw_error_code %d, maps to CRT "
+                  "error code %d",
+                  (void *)new_nw_socket,
+                  (void *)new_nw_socket->os_handle.nw_connection,
+                  nw_error_code,
+                  crt_error_code);
+
+              s_schedule_connection_state_changed_fn(
+                  new_socket, new_nw_socket, new_nw_socket->os_handle.nw_connection, state, crt_error_code);
+            });
 
-                AWS_LOGF_TRACE(
-                    AWS_LS_IO_SOCKET,
-                    "id=%p handle=%p: incoming connection connected to %s:%d, the incoming handle is %p",
-                    (void *)listener,
-                    listener->io_handle.data.handle,
-                    new_socket->remote_endpoint.address,
-                    new_socket->remote_endpoint.port,
-                    new_socket->io_handle.data.handle);
-
-                if (listener->accept_result_fn) {
-                    listener->accept_result_fn(listener, task_args->error_code, new_socket, task_args->user_data);
-                } else // The connection is not sent to user, clean it up. The nw_connection should be released in
-                       // socket clean up.
-                {
-                    AWS_FATAL_ASSERT("The listener accept_result_fn should not be NULL.");
-                }
-            }
-            s_unlock_base_socket(listener_nw_socket);
+        // this internal ref will be released when the connection canceled ( connection state changed to
+        // nw_connection_state_cancelled)
+        s_socket_acquire_internal_ref(new_nw_socket);
```

Comment:
Can we take this ref before we set the state changed callback?  What would happen if the connection got canceled in between setting the callback and this line?